### PR TITLE
Refactor afc pull method

### PR DESCRIPTION
**Pull request by guysalt**
bug fixes and improve of pull method in AfcService

### CODE REVIEW

**Review by doronz88 — CHANGES_REQUESTED** [review]
this PR breaks something. please test it first:
```shell
python3 -m pytest -vvv -x -k afc
```

**Review by doronz88 — CHANGES_REQUESTED** [review]
pull now fails on windows. probably because of misuse of path components

File: pymobiledevice3/services/afc.py [hunk 1]
```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
         super().__init__(lockdown, service_name)
         self.packet_num = 0
 
-    def pull(self, src: str, dst: str, callback=None):
+    def pull(self, src: str, dst: str, callback: Optional[callable] = None):
```
Comment (doronz88) [inline pymobiledevice3/services/afc.py]: use `typing.Callable`

File: pymobiledevice3/cli/afc.py [hunk 2]
```diff
@@ -24,11 +29,14 @@ def afc_shell(lockdown: LockdownClient):
 
 
 @afc.command('pull', cls=Command)
-@click.argument('remote_file', type=click.Path(exists=False))
-@click.argument('local_file', type=click.File('wb'))
-def afc_pull(lockdown: LockdownClient, remote_file, local_file):
-    """ pull remote file from /var/mobile/Media """
-    local_file.write(AfcService(lockdown=lockdown).get_file_contents(remote_file))
+@click.argument('remote_path', type=str)
```
Comment (doronz88) [inline pymobiledevice3/cli/afc.py:32]: remove the `type=str`

File: pyproject.toml [hunk 3]
```diff
@@ -28,7 +28,7 @@ classifiers = [
 dynamic = ["dependencies"]
 
 [project.optional-dependencies]
-test = ["pytest", "cmd2_ext_test"]
+test = ["pytest", "cmd2_ext_test", "gnureadline; platform_system != 'Windows'"]
```
Comment (doronz88) [inline pyproject.toml:31]: let's add this to the main requirements as well as the `pyreadline` so avoid annoying warning while using the CLI

File: tests/services/test_afc_shell.py [hunk 4]
```diff
@@ -28,19 +28,15 @@ def afc_shell(lockdown):
...
 
 
-def get_completions(line, part, app):
-    def get_line():
-        return line
+def get_completions(line: str, part: str, app: AfcShellTester) -> list[str]:
+    if sys.platform in ['win32', 'cygwin']:
+        import readline as testreadline
+    else:
+        import gnureadline as testreadline
```
Comment (doronz88) [inline tests/services/test_afc_shell.py]: reame to simply `readline`

**Review by doronz88 — CHANGES_REQUESTED** [review]

File: pymobiledevice3/services/afc.py [hunk 5]
```diff
@@ -772,7 +772,10 @@ def do_pull(self, args):
         def log(src, dst):
             self.poutput(f'{src} --> {dst}')
 
-        self.afc.pull(args.remote_path, args.local_path, callback=log, src_dir=self.curdir)
+        remote_path = str(pathlib.PurePosixPath(self.curdir).joinpath(args.remote_path))
+        if args.remote_path.endswith(posixpath.sep):
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py]: Add comment that explains the rational behind it (that pathlib removes the `/`)

File: pymobiledevice3/services/afc.py [hunk 6]
```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
         super().__init__(lockdown, service_name)
         self.packet_num = 0
 
-    def pull(self, src: str, dst: str, callback=None):
+    def pull(self, src: str, dst: str, callback: Optional[callable] = None):
         """ pull a file/directory from given path into a local file/directory """
-        if callback is not None:
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py:213]: Printing the source and the dest at start makes debugging easier

File: pymobiledevice3/services/afc.py [hunk 7]
```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
...
-
         src = self.resolve_path(src)
-
         if not self.isdir(src):
             # normal file
-            if os.path.isdir(dst):
-                dst = os.path.join(dst, os.path.basename(src))
-            with open(dst, 'wb') as f:
-                f.write(self.get_file_contents(src))
+            if dst.endswith(posixpath.sep):
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py]: `os.sep`?

File: pymobiledevice3/services/afc.py [hunk 8]
```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
...
+                dst_parent_dir = os.path.realpath(os.path.dirname(dst))
+            else:
+                dst_parent_dir = os.path.dirname(os.path.realpath(dst))
+            if not os.path.isdir(dst_parent_dir):  # check if the parent dir exist
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py]: `# check if the parent is dir`

File: pymobiledevice3/services/afc.py [hunk 9]
```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
...
+                raise NotADirectoryError(f'pull: directory {dst} does not exist')
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py]: `raise NotADirectoryError(f'pull: directory {dst_parent_dir} does not exist')`?

File: pymobiledevice3/services/afc.py [hunk 10]
```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
...
         src = self.resolve_path(src)
-
         if not self.isdir(src):
             # normal file
-            if os.path.isdir(dst):
-                dst = os.path.join(dst, os.path.basename(src))
-            with open(dst, 'wb') as f:
-                f.write(self.get_file_contents(src))
+            if dst.endswith(posixpath.sep):
+                dst_parent_dir = os.path.realpath(os.path.dirname(dst))
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py]:
```
dst_path = pathlib.Path(dst).expanduser().resolve()
if dst.endswith(os.sep):
    dst_path = dst_path / os.path.basename(src)
if not dst_path.parent.is_dir(): 
    raise NotADirectoryError(f'pull: directory {dst} does not exist')
dst_path.write_bytes(self.get_file_contents(src))
```

File: pymobiledevice3/services/afc.py (hunk 1 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: done

File: pymobiledevice3/cli/afc.py (hunk 2 above)
Comment (guysalt) [reply pymobiledevice3/cli/afc.py:32]: why?

File: pyproject.toml (hunk 3 above)
Comment (guysalt) [reply pyproject.toml:31]: you get pip error when you are trying to install pyreadline on unix os and gnureadline on windows os...

File: tests/services/test_afc_shell.py (hunk 4 above)
Comment (guysalt) [reply tests/services/test_afc_shell.py]: done

File: pymobiledevice3/services/afc.py (hunk 5 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: done

File: pymobiledevice3/services/afc.py (hunk 7 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: wow... i missed it, i think there is bug here on windows (also if i will put `os.path.sep`).

need to check it, thanks!

File: pymobiledevice3/services/afc.py (hunk 9 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: this is on purpose, act like `cp -r`

File: pymobiledevice3/services/afc.py (hunk 8 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: done

File: pymobiledevice3/services/afc.py [hunk 11]
```diff
@@ -208,36 +208,43 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
...
-            dst_path.mkdir(parents=True, exist_ok=True)
-
-            for filename in self.listdir(src):
-                src_filename = posixpath.join(src, filename)
-                dst_filename = dst_path / filename
+            if os.path.isfile(dst):
+                raise NotADirectoryError('dst is not a directory (while src is a directory)')
 
-                src_filename = self.resolve_path(src_filename)
+            if os.path.isdir(dst):
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py]: exists?

File: pymobiledevice3/services/afc.py (hunk 10 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: this block have couple of bugs, like pull src file to existing folder without trailing slash.

but i took inspiration from this Implementation:
```python
if dst.endswith((os.path.sep, posixpath.sep)) and not dst_path.is_dir():
    raise NotADirectoryError(f'pull: directory {dst_path} does not exist')

if dst_path.is_dir():
    dst_path = dst_path.joinpath(os.path.basename(src))

if not dst_path.parent.is_dir():
    raise FileNotFoundError(f'pull: {dst_path}: No such file or directory')
```

also, in this Implementation you distinguish between the errors  (and need to add tests for this).
(:

File: pymobiledevice3/services/afc.py (hunk 9 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: refactor this area

File: pymobiledevice3/services/afc.py (hunk 10 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: the last part (`if not dst_path.parent.is_dir():`) move to be in the start of the function :)

File: pymobiledevice3/services/afc.py (hunk 11 above)
Comment (guysalt) [reply pymobiledevice3/services/afc.py]: in this line it doesn't mean the same? as we check already dst is not file

File: pymobiledevice3/services/afc.py (hunk 6 above)
Comment (matan1008) [reply pymobiledevice3/services/afc.py:213]: @guysalt

File: pymobiledevice3/services/afc.py [hunk 12]
```diff
@@ -208,36 +208,48 @@ def __init__(self, lockdown: LockdownClient, service_name='com.apple.afc'):
...
-                src_filename = self.resolve_path(src_filename)
+            if dst_path.exists() and not src.endswith(posixpath.sep):
+                dst_path /= os.path.basename(src)
 
-                if self.isdir(src_filename):
-                    dst_filename.mkdir(exist_ok=True)
-                    self.pull(src_filename, str(dst_path), callback=callback)
-                    continue
+            try:
+                dst_path.mkdir()
```
Comment (matan1008) [inline pymobiledevice3/services/afc.py:243]: exists_ok=True

**Pull request closed without merging**